sound/soc/sof/ipc.c

```diff
@@ -386,16 +386,30 @@ EXPORT_SYMBOL(sof_ipc_drop_all);
 int snd_sof_ipc_reply(struct snd_sof_dev *sdev, u32 msg_id)
 {
 	struct snd_sof_ipc_msg *msg;
+	unsigned long flags;
+
+	/*
+	 * Protect against a theoretical race with ipc_tx_next_msg(): if the DSP
+	 * is fast enough to receive an IPC message, reply to it, and the host
+	 * interrupt processing calls this function on a different core from the
+	 * one, where the sending is taking place, the message might not yet be
+	 * on the .reply_list and sof_ipc_reply_find_msg() will return 0.
+	 */
+	spin_lock_irqsave(&sdev->ipc_lock, flags);
 
 	msg = sof_ipc_reply_find_msg(sdev->ipc, msg_id);
 	if (!msg) {
+		spin_unlock_irqrestore(&sdev->ipc_lock, flags);
 		dev_err(sdev->dev, "error: can't find message header 0x%x",
 			msg_id);
 		return -EINVAL;
 	}
 
 	/* wake up and return the error if we have waiters on this message ? */
 	sof_ipc_tx_msg_reply_complete(sdev->ipc, msg);
```

Comment:
@lyakh do we need this function? I think it makes sense to just set ipc_complete and wakeUp here. That way it is easier to tell that it is in the locked context.

Reply:
I was on and off about this, in general I'd tend to agree, that it's easier to read with that function inlined, but I decided, that it would add just one more trivial patch to this already rather intense patch series, so, I decided to leave it out for now, we can do it later.